src/fibex.rs

```diff
@@ -0,0 +1,1642 @@
+                    model.types.append(&mut service_types);
+                }
+                FibexEvent::DatatypeStart { id, datatype_type } => {
+                    if datatype_type == "fx:COMMON-DATATYPE-TYPE" {
+                        model.types.push(Rc::new(RefCell::new(parse_common_datatype(
+                            &mut reader,
+                            id,
+                        )?)));
+                    } else if datatype_type == "fx:COMPLEX-DATATYPE-TYPE" {
+                        model
+                            .types
+                            .push(Rc::new(RefCell::new(parse_complex_datatype(
+                                &mut reader,
+                                id,
+                            )?)));
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        model.pack(strict)?;
+
+        Ok(model)
+    }
+
+    fn parse_service_interface<R: Read>(
+        reader: &mut FibexReader<BufReader<R>>,
+        id: String,
+    ) -> Result<
+        (
+            FibexServiceInterface,
+            Vec<Rc<RefCell<FibexTypeDeclaration>>>,
+        ),
+        FibexError,
+    > {
+        let mut service_name: Option<String> = None;
+        let mut service_id: Option<usize> = None;
+        let mut major_version: Option<usize> = None;
+        let mut minor_version: Option<usize> = None;
+        let mut methods: Vec<FibexServiceMethod> = Vec::new();
+        let mut types: Vec<Rc<RefCell<FibexTypeDeclaration>>> = Vec::new();
+
+        loop {
+            match reader.read_fibex()? {
+                FibexEvent::ShortName { name } => {
+                    service_name = Some(name);
+                }
+                FibexEvent::ServiceIdentifier { identifier } => {
+                    service_id = Some(parse_number(&id, "service_id", &identifier)?);
+                }
+                FibexEvent::MajorVersion { version } => {
+                    major_version = Some(parse_number(&id, "major_version", &version)?);
+                }
+                FibexEvent::MinorVersion { version } => {
+                    minor_version = Some(parse_number(&id, "minor_version", &version)?);
+                }
+                FibexEvent::MethodStart { id } => {
+                    let (method, mut method_types) = parse_service_method(reader, id)?;
+                    methods.push(method);
+                    types.append(&mut method_types);
+                }
+                FibexEvent::FieldStart { id } => {
+                    let (mut field_methods, field) = parse_service_field(reader, id)?;
+                    methods.append(&mut field_methods);
+                    types.push(field);
+                }
+                FibexEvent::ServiceEnd {} => {
+                    return Ok((
+                        FibexServiceInterface {
+                            id: id.clone(),
+                            name: move_value(&id, "service_name", service_name)?,
```

Comment:
no need to call a function to turn an option into a result. You can use
```rust
                            name: service_name.ok_or_else(|| {
                                FibexError::Model(format!("Missing service_name at {}", id,))
                            })?,
```
and if you prefer a function, do not call it `move` since it is not about moving a value...just about converting an option into a `Result`

Reply:
Das ist interessant,- ich dachte eigentlich, dass der Wert von dem lokalen Option<T> service_name tatsächlich bei der Zuweisung an den return von FibexServiceInterface::name ge-moved wird?!

let mut service_name: Option<String> = None;
..
return Ok((
  FibexServiceInterface {
    ..
    name: move_value(&id, "service_name", service_name)?
..
));

Eine Funktion hab ich anstelle dem ok_or_else genommen um den boilerplate code beim Zusammenbauen der Model-Elemente aus den zwischenzeitlich (hoffentlich => Option<T>) geparsten Daten möglichst gering/flach zu halten..

Reply:
done